src/System.Linq.Queryable/src/System/Linq/EnumerableQuery.cs

```diff
@@ -135,7 +135,10 @@ private IEnumerator<T> GetEnumerator()
                 EnumerableRewriter rewriter = new EnumerableRewriter();
                 Expression body = rewriter.Visit(_expression);
                 Expression<Func<IEnumerable<T>>> f = Expression.Lambda<Func<IEnumerable<T>>>(body, (IEnumerable<ParameterExpression>)null);
-                _enumerable = f.Compile()();
+                IEnumerable<T> enumerable = f.Compile()();
+                if (enumerable == this)
+                    throw new InvalidOperationException();
```

Comment:
This should probably have a descriptive message associated with it (unless we don't do that throughout this project?)